### DISCUSSION TITLE
Multi-instance Support on NATIVE Posix Command Line Argument Access

### DISCUSSION THREAD

**Discussion by bob2oneil** [zephyrproject-rtos/zephyr#35803] [Q&A] [answered]
I want to run multiple instances of a Zephyr application on a Linux box running under the NATIVE POSIX 64 environment.

Each instance would need to create two virtual Ethernet interfaces with different IP/MAC addresses for each of these 2 virtual TAP interface runtime.

I would prefer this to occur at boot, but it is not clear how each instance will know the number of other instances present.

Under embedded Linux I would use something like "pidof" to determine the count of running instances.

I could also use something like global shared memory or a file to keep track of the number of running instances.  But I do not want to have this based on storing content to the littlefs - but it is an option of nothing else is discovered that is better.  Since under NATIVE POSIX, the littlefs seems to write to a single file as FLASH.BIN, presumably that file would need to be used by all instances.

Notionally, the 1st instance would get assigned a node id of 1, would use this information to create unique IP and MAC addresses for each virtual Ethernet interface.  These interfaces need to act like promiscuous sockets with a larger MTU of 9k.

If command line parameter were available, the node id could be assigned via a command line argument.

How would you fine folks solve this issue?

**Comment by nashif** [comment, marked as answer]
native_posix does support command line arguments, and it can be extended as well. Search for `NATIVE_TASK` which can call a hook to add new command line options.

**Reply by bob2oneil** [reply]
nashif, thanks for the input.  I would like to capture the command line arguments, presumably if a callback is registered PRE_BOOT_1, the callback receives the traditional argc, argv?  Where would I find more information on this use to determine what is provided to the callback?

**Reply by yashi** [reply]
`native_add_command_line_opts()` extends and add new command line options.  You can also check `struct args_struct_t` for `argc` and `argv` handling.

**Reply by bob2oneil** [reply]
Thanks much yashi and nashif.

**Reply by yashi** [reply]
@bob2oneil, please mark it "answered" to help others find it.  Thanks.

**Reply by aescolar** [reply]
@bob2oneil You can check as examples https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/posix/native_posix/timer_model.c#L546 
or https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/flash/flash_simulator.c#L434

**Reply by bob2oneil** [reply]
Thanks aescolar and the rest of the team.  I am unable to get this to build in a sample application.  I get a compiler error resolving the definition of the callback to the NATIVE_TASK define.  I do have both soc.h (defines NATIVE_TASK) and cmdline.h added for the POSIX architecture.  If I cast the value to a "void *", other build errors show themselves.

Any ideas on this build issue?  Is this a C++ issue with the definition of the callback?

Attached is the main.cpp.

Compiler build errors are as follows:
```

[231/239] Building CXX object CMakeFiles/app.dir/src/main.cpp.obj
FAILED: CMakeFiles/app.dir/src/main.cpp.obj 
ccache /usr/bin/gcc  -DBUILD_VERSION=v2.6.0-rc3 -DKERNEL -DMBEDTLS_CONFIG_FILE=\"config-tls-generic.h\" -D_FORTIFY_SOURCE=2 -D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED -D__ZEPHYR__=1 -I/home/robocom/zephyrproject/zephyr/include -Izephyr/include/generated -I/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock -I/home/robocom/zephyrproject/zephyr/boards/posix/native_posix -I/home/robocom/zephyrproject/zephyr/lib/util/fnmatch/. -I/home/robocom/zephyrproject/zephyr/subsys/net/l2 -I/home/robocom/zephyrproject/modules/crypto/mbedtls/mbedtls/include -I/home/robocom/zephyrproject/zephyr/modules/mbedtls/configs -Og -fcheck-new -fno-exceptions -fno-rtti -imacros /home/robocom/RCSUS/promisc/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -fdiagnostics-color=always -imacros /home/robocom/zephyrproject/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-address-of-packed-member -Wno-unused-but-set-variable -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/robocom/RCSUS/promisc=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/robocom/zephyrproject/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/robocom/zephyrproject=WEST_TOPDIR -ffunction-sections -fdata-sections -m64 -fPIC -include /home/robocom/zephyrproject/zephyr/arch/posix/include/posix_cheats.h -fno-freestanding -MD -MT CMakeFiles/app.dir/src/main.cpp.obj -MF CMakeFiles/app.dir/src/main.cpp.obj.d -o CMakeFiles/app.dir/src/main.cpp.obj -c ../src/main.cpp
In file included from ../src/main.cpp:12:
/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock/soc.h:46:14: warning: unnecessary parentheses in declaration of ‘__native_task_add_user_arg’ [-Wparentheses]
   46 |  static void (*_CONCAT(__native_task_, fn)) __used \
      |              ^
../src/main.cpp:347:1: note: in expansion of macro ‘NATIVE_TASK’
  347 | NATIVE_TASK(add_user_arg, PRE_BOOT_1, 1);
      | ^~~~~~~~~~~
../src/main.cpp:347:13: error: invalid conversion from ‘void (*)()’ to ‘void*’ [-fpermissive]
  347 | NATIVE_TASK(add_user_arg, PRE_BOOT_1, 1);
      |             ^~~~~~~~~~~~
      |             |
      |             void (*)()
/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock/soc.h:48:4: note: in definition of macro ‘NATIVE_TASK’
   48 |  = fn
      |    ^~
[232/239] Linking C static library zephyr/kernel/libkernel.a
ninja: build stopped: subcommand failed.

```

**Reply by bob2oneil** [reply]
The expansion of the NATIVE_TASK macro under C++ will not resolve properly, even with the callback in a separate 'C' file with the extern 'C' keywords.

```
In file included from ../src/main.cpp:12:
/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock/soc.h:46:14: warning: unnecessary parentheses in declaration of ‘__native_task_add_user_arg’ [-Wparentheses]
   46 |  static void (*_CONCAT(__native_task_, fn)) __used \
      |              ^
../src/main.cpp:369:1: note: in expansion of macro ‘NATIVE_TASK’
  369 | NATIVE_TASK(add_user_arg, PRE_BOOT_1, 1);
      | ^~~~~~~~~~~
../src/main.cpp:369:13: error: invalid conversion from ‘void (*)()’ to ‘void*’ [-fpermissive]
  369 | NATIVE_TASK(add_user_arg, PRE_BOOT_1, 1);
      |             ^~~~~~~~~~~~
      |             |
      |             void (*)()
/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock/soc.h:48:4: note: in definition of macro ‘NATIVE_TASK’
   48 |  = fn
      |    ^~
[232/240] Linking C static library modules/mbedtls/libmodules__mbedtls.a
```

**Comment by bob2oneil** [comment]
Thanks to the community for providing a response, I believe I have a direction on this issue.

**Comment by nashif** [comment]
@aescolar  maybe worth documenting this somewhere if not covered already.

**Reply by aescolar** [reply]
@nashif Yep. Good point.

**Comment by bob2oneil** [comment]
[main.zip](https://github.com/zephyrproject-rtos/zephyr/files/6627334/main.zip)

**Reply by bob2oneil** [reply]
When callback is cast to a (void *), the following build errors unsue.
```

main.cpp.obj -c ../src/main.cpp
In file included from /home/robocom/zephyrproject/zephyr/include/toolchain/gcc.h:69,
                 from /home/robocom/zephyrproject/zephyr/include/toolchain.h:43,
                 from /home/robocom/zephyrproject/zephyr/include/kernel_includes.h:19,
                 from /home/robocom/zephyrproject/zephyr/include/kernel.h:17,
                 from /home/robocom/zephyrproject/zephyr/include/zephyr.h:18,
                 from ../src/main.cpp:1:
/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock/soc.h:46:24: error: pasting "__native_task_" and "(" does not give a valid preprocessing token
   46 |  static void (*_CONCAT(__native_task_, fn)) __used \
      |                        ^~~~~~~~~~~~~~
/home/robocom/zephyrproject/zephyr/include/toolchain/common.h:124:26: note: in definition of macro ‘_DO_CONCAT’
  124 | #define _DO_CONCAT(x, y) x ## y
      |                          ^
/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock/soc.h:46:16: note: in expansion of macro ‘_CONCAT’
   46 |  static void (*_CONCAT(__native_task_, fn)) __used \
      |                ^~~~~~~
../src/main.cpp:347:1: note: in expansion of macro ‘NATIVE_TASK’
  347 | NATIVE_TASK((void *)add_user_arg, PRE_BOOT_1, 1);
      | ^~~~~~~~~~~
../src/main.cpp:347:21: error: expected ‘)’ before ‘add_user_arg’
  347 | NATIVE_TASK((void *)add_user_arg, PRE_BOOT_1, 1);
      |                     ^~~~~~~~~~~~
/home/robocom/zephyrproject/zephyr/include/toolchain/common.h:124:31: note: in definition of macro ‘_DO_CONCAT’
  124 | #define _DO_CONCAT(x, y) x ## y
      |                               ^
/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock/soc.h:46:16: note: in expansion of macro ‘_CONCAT’
   46 |  static void (*_CONCAT(__native_task_, fn)) __used \
      |                ^~~~~~~
../src/main.cpp:347:1: note: in expansion of macro ‘NATIVE_TASK’
  347 | NATIVE_TASK((void *)add_user_arg, PRE_BOOT_1, 1);
      | ^~~~~~~~~~~
In file included from ../src/main.cpp:12:
/home/robocom/zephyrproject/zephyr/soc/posix/inf_clock/soc.h:46:14: note: to match this ‘(’
   46 |  static void (*_CONCAT(__native_task_, fn)) __used \
      |              ^
../src/main.cpp:347:1: note: in expansion of macro ‘NATIVE_TASK’
  347 | NATIVE_TASK((void *)add_user_arg, PRE_BOOT_1, 1);
      | ^~~~~~~~~~~
../src/main.cpp:280:13: warning: ‘void add_user_arg()’ defined but not used [-Wunused-function]
  280 | static void add_user_arg(void)
      |             ^
```

**Reply by bob2oneil** [reply]
Appears to be a C++ issue with the callback as file is main.cpp.  I tried putting the callback into a separate C file with an extern "C" declaration, and the problem continued with the same error resolving the NATIVE_TASK macro expansion in C++.  The application will be written in C++, so I need the main.cpp to likewise be C++ to build.

**Reply by aescolar** [reply]
did you put both the callback, the `NATIVE_TASK()` line and `add_user_arg` in its own separate C file?
(anyhow this is a mingling C and C++ issue, not really related to the first ;) )

**Reply by bob2oneil** [reply]
Really good advice Alberto, I sort of assumed the NATIVE_TASK had to reside with main(), perhaps for no real good reason.  Moving all of this to 'C' file, now I am able to build.  This might be something other folks will wander into again, that is getting command line support for C++ based implementations in Zephyr, or more broadly mixing Zephyr C code with C++ where certain builds require 'C'.
